graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/conversion/FieldValueType.java

```diff
@@ -0,0 +1,109 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ValueNode;
+import com.google.common.collect.ImmutableList;
+import org.graylog.plugins.pipelineprocessor.EvaluationContext;
+import org.graylog.plugins.pipelineprocessor.ast.functions.AbstractFunction;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionArgs;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionDescriptor;
+import org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor;
+import org.graylog.plugins.pipelineprocessor.rulebuilder.RuleBuilderFunctionGroup;
+import org.graylog2.plugin.Message;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor.type;
+
+public class FieldValueType extends AbstractFunction<String> {
+    public static final String NAME = "field_value_type";
+    public static final String FIELD = "field";
+
+    public static final String BOOLEAN = "boolean";
+    public static final String STRING = "string";
+    public static final String DOUBLE = "double";
+    public static final String LONG = "long";
+    public static final String LIST = "list";
+    public static final String MAP = "map";
+    public static final String VALUE_NODE = "value_node";
+    public static final String ARRAY_NODE = "array_node";
+    public static final String OBJECT_NODE = "object_node";
+
+    private final ParameterDescriptor<String, String> fieldParam;
+    private final ParameterDescriptor<Message, Message> messageParam;
+
+    public FieldValueType() {
+        fieldParam = ParameterDescriptor.string(FIELD).description("The field to get the value type of").build();
+        messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").ruleBuilderVariable().build();
+    }
+
+    @Override
+    public String evaluate(FunctionArgs args, EvaluationContext context) {
+        final String field = fieldParam.required(args, context);
+        final Message message = messageParam.optional(args, context).orElse(context.currentMessage());
+        final var value = message.getField(field);
+
+        if (value instanceof Boolean) {
+            return BOOLEAN;
+        }
+        if (value instanceof String) {
+            return STRING;
+        }
+        if (value instanceof Double) {
+            return DOUBLE;
+        }
+        if (value instanceof Long) {
+            return LONG;
+        }
+        if (value instanceof List) {
+            return LIST;
+        }
+        if (value instanceof Map) {
+            return MAP;
+        }
+        if (value instanceof ValueNode) {
+            return VALUE_NODE;
+        }
+        if (value instanceof ArrayNode) {
+            return ARRAY_NODE;
+        }
+        if (value instanceof ObjectNode) {
+            return OBJECT_NODE;
+        }
```

Comment:
Nit: It seems like these types could utilize an enum class to define their constants and corresponding string values. Within that enum class, I think the `valueOf()` method could be overridden to perform the instance of check and return the corresponding `value`, potentially without needing an `instanceof` if chain or case statement. Then, in the function, I think just an `EnumClass.valueOf(value)` could be used. It might be worth checking if that is a slight improvement.

Reply:
Here's one similar example: https://github.com/Graylog2/graylog2-server/blob/2a66c130fef1f95702ed212a6ba7fa258c835351/graylog2-server/src/main/java/org/graylog2/indexer/datastream/policy/actions/TimesUnit.java#L25

It looks like `valueOf` cannot be overridden, but maybe an additional method is an option.

Reply:
Nice! Yeah good call, I think that'll be a lot cleaner so will take a stab at it. Thanks for the example!